Add new _gasPrice reserved parameter

dcroote opened the pull request:
Allows gasPrice to be specified by a requester.

Closes #1492

dcroote commented on packages/airnode-adapter/src/types.ts:
```diff
@@ -62,7 +62,8 @@ export type BaseResponseType = typeof baseResponseTypes[number];
 export type ResponseType = string;
 
-export interface ReservedParameters {
+// Reserved parameters specific for response processing i.e. excludes _gasPrice
+export interface ResponseReservedParameters {
```
I renamed `ReservedParameters` to `ResponseReservedParameters` to distinguish those that are relevant to response processing (`_type`, `_path`, and `_times`) from those that are not i.e. `_gasPrice` and (coming soon) `_minConfirmations`

dcroote commented on packages/airnode-node/src/types.ts:
```diff
@@ -229,6 +229,7 @@ export type ApiCallResponse = RegularApiCallResponse | HttpGatewayApiCallRespons
 export interface RegularApiCallSuccessResponse {
   success: true;
   data: { encodedValue: string; signature: string };
+  gasPriceOverride?: string;
```
I spent a lot of time trying to find a way of retaining the requester-specified `_gasPrice` from the where API call response processing and reserved parameter parsing occurs for use where the `gasTarget` created and employed in submitting transactions. Obviously reserved parameters weren't exactly designed / implemented for more than response processing, but adding this property was the simplest solution I came up with.

dcroote replied:
Do you mean something like
```ts
export interface RegularApiCallSuccessResponse {
  success: true;
  data: { encodedValue: string; signature: string };
  reservedParameterOverrides?: {
    _gasPrice?: string;
    _minConfirmations?: string;
  }
}
```

amarthadan replied:
Yes.

dcroote commented on packages/airnode-node/src/api/index.ts:
```diff
@@ -235,10 +235,11 @@ export async function processSuccessfulApiCall(
-      goPostProcessApiSpecifications.data,
-      reservedParameters as adapter.ReservedParameters
-    )
+    adapter.extractAndEncodeResponse(goPostProcessApiSpecifications.data, {
+      _type,
+      _path,
+      _times,
+    } as adapter.ResponseReservedParameters)
```
Similarly here, trying to be explicit about which reserved parameters are relevant to response processing.

dcroote commented on packages/airnode-node/src/types.ts:
```diff
@@ -244,6 +245,7 @@ export interface HttpSignedDataApiCallSuccessResponse {
 export interface ApiCallErrorResponse {
   success: false;
   errorMessage: string;
+  gasPriceOverride?: string;
```
Needed to add it here in order to match `RegularApiCallSuccessResponse` (satisfy type constraints)

amarthadan reviewed (commented):
LGTM, just a minor note. I want to test it first before approving.

amarthadan reviewed (changes requested):
Have you tested this? Because I did and I'm pretty sure it couldn't have worked :smile:
It took me quite some time but I found out what the problem is and why the value is ignored. The problem is here
https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/coordinator/calls/disaggregation.ts#L41
The response there is reconstructed but only the fields `data` and `success` are added. That was enough before but now we need to also add the `reservedParameterOverrides` field.
TS didn't complain because the field is optional anyway. But the set `_gasPrice` value was ignored.

Also, even though the value correctly overrides what's returned from the gas price oracle, the logs are quite confusing. They say how is the value chosen by the oracle but they don't mention anything about it being overridden later.

I'd also suggest writing tests using these reserved parameter overrides so that we won't run into something similar again.

amarthadan reviewed (approved):
LGTM

dcroote commented:
Oof 🤦 😞 well an apology is in order @amarthadan. I thought I had covered it through the tests I updated, but clearly you were right in this not working due to reconstruction. I added the E2E and confirmed it's working as expected locally. Sorry you ended up spending time debugging.
